components/script/stylesheet_loader.rs

```diff
@@ -0,0 +1,254 @@
+            successful = metadata.status.map_or(false, |(code, _)| code == 200);
+        }
+
+        if let Some(ref link) = elem.downcast::<HTMLLinkElement>() {
+            if link.parser_inserted() {
+                document.decrement_script_blocking_stylesheet_count();
+            }
+        } else if let Some(ref style) = elem.downcast::<HTMLStyleElement>() {
+            if style.parser_inserted() {
+                document.decrement_script_blocking_stylesheet_count();
+            }
+        } else {
+            unreachable!(
+                "Stylesheet loads can only be triggered by <link> or <style> elements!");
+        }
+
+        let url = self.source.url();
+        document.finish_load(LoadType::Stylesheet(url));
+
+        if let Some(ref link) = elem.downcast::<HTMLLinkElement>() {
+            if let Some(any_failed) = link.load_finished(successful) {
+                let event = if any_failed { atom!("error") } else { atom!("load") };
+                link.upcast::<EventTarget>().fire_event(event);
+            }
+        } else if let Some(ref style) = elem.downcast::<HTMLStyleElement>() {
+            if let Some(any_failed) = style.load_finished(successful) {
+                let event = if any_failed { atom!("error") } else { atom!("load") };
+                style.upcast::<EventTarget>().fire_event(event);
+            }
+        } else {
+            unreachable!(
+                "Stylesheet loads can only be triggered by <link> or <style> elements!");
+        }
+    }
+}
+
+pub struct StylesheetLoader<'a> {
+    elem: &'a HTMLElement,
+}
+
+impl<'a> StylesheetLoader<'a> {
+    pub fn for_element(element: &'a HTMLElement) -> Self {
+        StylesheetLoader {
+            elem: element,
+        }
+    }
+}
+
+impl<'a> StylesheetLoader<'a> {
+    pub fn load(&self, source: StylesheetContextSource) {
+        let url = source.url();
+        let context = Arc::new(Mutex::new(StylesheetContext {
+            elem: Trusted::new(&*self.elem),
+            source: source,
+            metadata: None,
+            data: vec![],
+        }));
+
+        let document = document_from_node(self.elem);
+
+        let (action_sender, action_receiver) = ipc::channel().unwrap();
+        let listener = NetworkListener {
+            context: context,
+            task_source: document.window().networking_task_source(),
+            wrapper: Some(document.window().get_runnable_wrapper())
+        };
+        ROUTER.add_route(action_receiver.to_opaque(), box move |message| {
+            listener.notify_fetch(message.to().unwrap());
+        });
+
+
+        let mut referrer_policy = document.get_referrer_policy();
+        if let Some(ref link) = self.elem.downcast::<HTMLLinkElement>() {
+            link.increment_pending_loads_count();
+            if link.parser_inserted() {
+                document.increment_script_blocking_stylesheet_count();
+            }
+            if link.RelList().Contains("noreferrer".into()) {
```

Comment:
Let's pass this in as an argument, and have a trait that `HTMLLinkElement` and `HTMLStyleElement` both implement, instead of duplicating all the code.